bundles/org.openhab.binding.pilight/src/main/java/org/openhab/binding/pilight/internal/discovery/PilightDeviceDiscoveryService.java

```diff
@@ -0,0 +1,215 @@
+import org.openhab.binding.pilight.internal.PilightHandlerFactory;
+import org.openhab.binding.pilight.internal.dto.Config;
+import org.openhab.binding.pilight.internal.dto.DeviceType;
+import org.openhab.binding.pilight.internal.dto.Status;
+import org.openhab.binding.pilight.internal.handler.PilightBridgeHandler;
+import org.openhab.core.config.discovery.AbstractDiscoveryService;
+import org.openhab.core.config.discovery.DiscoveryResult;
+import org.openhab.core.config.discovery.DiscoveryResultBuilder;
+import org.openhab.core.config.discovery.DiscoveryService;
+import org.openhab.core.thing.ThingTypeUID;
+import org.openhab.core.thing.ThingUID;
+import org.openhab.core.thing.binding.ThingHandler;
+import org.openhab.core.thing.binding.ThingHandlerService;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link PilightDeviceDiscoveryService} discovers pilight devices after a bridge thing has been created and
+ * connected to the pilight daemon. Things are discovered periodically in the background or after a manual trigger.
+ *
+ * @author Niklas Dörfler - Initial contribution
+ */
+@NonNullByDefault
+@Component(service = DiscoveryService.class, immediate = true, configurationPid = "discovery.pilight")
+public class PilightDeviceDiscoveryService extends AbstractDiscoveryService implements ThingHandlerService {
+
+    private static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = PilightHandlerFactory.SUPPORTED_THING_TYPES_UIDS;
+
+    private static final int AUTODISCOVERY_SEARCH_TIME_SEC = 10;
+    private static final int AUTODISCOVERY_BACKGROUND_SEARCH_INTERVAL_SEC = 60 * 10;
+
+    private final Logger logger = LoggerFactory.getLogger(PilightDeviceDiscoveryService.class);
+
+    private @Nullable PilightBridgeHandler pilightBridgeHandler;
+    private @Nullable ThingUID bridgeUID;
+
+    private @Nullable ScheduledFuture<?> backgroundDiscoveryJob;
+    private CompletableFuture<Config> configFuture;
+    private CompletableFuture<List<Status>> statusFuture;
+
+    public PilightDeviceDiscoveryService() {
+        super(SUPPORTED_THING_TYPES_UIDS, AUTODISCOVERY_SEARCH_TIME_SEC);
+        configFuture = new CompletableFuture<>();
+        statusFuture = new CompletableFuture<>();
+    }
+
+    @Override
+    protected void startScan() {
+        if (pilightBridgeHandler != null) {
+            configFuture = new CompletableFuture<>();
+            statusFuture = new CompletableFuture<>();
+
+            configFuture.thenAcceptBoth(statusFuture, (config, allStatus) -> {
+                removeOlderResults(getTimestampOfLastScan(), bridgeUID);
+                config.getDevices().forEach((deviceId, device) -> {
+                    if (this.pilightBridgeHandler != null) {
+                        final Optional<Status> status = allStatus.stream()
+                                .filter(s -> deviceId.equals(s.getDevices().get(0))).findFirst();
```

Comment:
Would this be more correct?
```suggestion
                                .filter(s -> s.getDevices().contains(deviceId)).findFirst();
```